File: solutions/cpp/0022.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ret;
        helper("", n, n, ret);
        return ret;
    }

private:
    void helper(string str, int l, int r, vector<string>& ret) {
        if (l == 0 && r == 0) ret.push_back(str);
        if (l > 0) helper(str + '(', l - 1, r, ret);
        if (l < r) helper(str + ')', l, r - 1, ret);
    }
};

File: solutions/python3/0022.py
class Solution:
    def generateParenthesis(self, n):
        ret = []
        self.helper('', n, n, ret)
        return ret

    def helper(self, str, l, r, ret):
        if l == 0 and r == 0:
            ret.append(str)
        if l > 0:
            self.helper(str + '(', l - 1, r, ret)
        if l < r:
            self.helper(str + ')', l, r - 1, ret)